src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs

```diff
@@ -69,47 +73,115 @@ IAsyncServiceProvider serviceProvider
             ISolutionRestoreWorker restoreWorker,
             ILogger logger,
-            Microsoft.VisualStudio.Threading.AsyncLazy<IVsSolution2> vsSolution2)
+            Microsoft.VisualStudio.Threading.AsyncLazy<IVsSolution2> vsSolution2,
+            INuGetTelemetryProvider telemetryProvider)
         {
             _projectSystemCache = projectSystemCache ?? throw new ArgumentNullException(nameof(projectSystemCache));
             _restoreWorker = restoreWorker ?? throw new ArgumentNullException(nameof(restoreWorker));
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
             _vsSolution2 = vsSolution2 ?? throw new ArgumentNullException(nameof(vsSolution2));
+            _telemetryProvider = telemetryProvider ?? throw new ArgumentNullException(nameof(telemetryProvider));
         }
 
-        public Task<bool> CurrentRestoreOperation => _restoreWorker.CurrentRestoreOperation;
+        Task<bool> IVsSolutionRestoreService.CurrentRestoreOperation
+        {
+            get
+            {
+                const string eventName = nameof(IVsSolutionRestoreService) + "." + nameof(IVsSolutionRestoreService.CurrentRestoreOperation);
+                NuGetExtensibilityEtw.EventSource.Write(eventName, NuGetExtensibilityEtw.InfoEventOptions);
+                return _restoreWorker.CurrentRestoreOperation;
+            }
+        }
 
-        public async Task<bool> NominateProjectAsync(string projectUniqueName, CancellationToken token)
+        Task<bool> IVsSolutionRestoreService3.CurrentRestoreOperation
         {
-            Assumes.NotNullOrEmpty(projectUniqueName);
+            get
+            {
+                const string eventName = nameof(IVsSolutionRestoreService) + "." + nameof(IVsSolutionRestoreService3.CurrentRestoreOperation);
+                NuGetExtensibilityEtw.EventSource.Write(eventName, NuGetExtensibilityEtw.InfoEventOptions);
+                return _restoreWorker.CurrentRestoreOperation;
+            }
+        }
 
-            ProjectNames projectNames = await GetProjectNamesAsync(projectUniqueName, token);
-            var dgSpec = new DependencyGraphSpec();
-            var packageSpec = new PackageSpec()
+        public async Task<bool> NominateProjectAsync(string projectUniqueName, CancellationToken token)
+        {
+            const string eventName = nameof(IVsSolutionRestoreService2) + "." + nameof(NominateProjectAsync);
+            var eventData = new NominateProjectAsyncEventData()
             {
-                Name = projectUniqueName
+                ProjectUniqueName = projectUniqueName
             };
-            dgSpec.AddProject(packageSpec);
-            dgSpec.AddRestore(packageSpec.Name);
-            _projectSystemCache.AddProjectRestoreInfo(projectNames, dgSpec, new List<IAssetsLogMessage>());
+            NuGetExtensibilityEtw.EventSource.Write(eventName, NuGetExtensibilityEtw.StartEventOptions, eventData);
```

Comment:
As per comment from the `dotnet/runtime` repo, `Write` method is not as efficient at runtime as `WriteEvent` methods.

https://github.com/dotnet/runtime/blob/a891aed8e03a225d8b31bbea0ae1a86972adc819/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L33-L39

May be this is the reason the below guidance states `Do specify explicit event ID values using EventAttribute for the ETW event methods. `

https://github.com/Microsoft/dotnet-samples/blob/master/Microsoft.Diagnostics.Tracing/EventSource/docs/EventSource.md#event-source-design-guidelines

I remember that during our offline conversation you have mentioned that `Write` methods are used instead of `WriteEvent` because `Write` methods accept a `string` parameter for event name.  I am fine with the current implementation but wanted to share this information in case if it helps.

Reply:
As discussed, it would require an additional place where all APIs would need to be duplicated, since we'd need one or two methods (start and stop) in an EventSource class for every API we have. So, without a source generator to write most of the code for us, it would be a lot more boilerplate than this PR already has.

Given the linked comments also say "these are not quite as efficient at runtime since you have at least a hash table lookup on every event write.", I don't believe that a hash table lookup is very expensive, especially compared to how much work APIs like GetInstalledPackagesAsync, InstallPackage, UninstallPackage, etc are. I think the `Write<T>` overload for writing event data might be somewhat more expensive, since it will need to use reflection to figure out the ETW event metadata. However, there would be a lot of effort required to change this just for perf reasons, so I think the famous "premature optimization is the root of all evil" saying applies here. Unless with have performance traces that show that our EW code has impact on VS performance, the effort to implement a higher performance implementation is probably not worthwhile.